### ISSUE TITLE
Migration attempt results in:  Neither DSN nor SERVER keyword supplied (0)

### ISSUE THREAD

**Issue by clabornd** [sqlalchemy/sqlalchemy#5332]
Using a connection string of the form:

```
params = urllib.parse.quote_plus \
("Driver={ODBC Driver 17 for SQL Server};Server=servername;Database=dbname;Uid=username;Pwd=mypword;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;")
conn_str = 'mssql+pyodbc:///?odbc_connect={}'.format(params)

class Config(object):
    SQLALCHEMY_DATABASE_URI = conn_str
```

flask migrate -m "msg" results in:

Traceback (most recent call last):
  File "/anaconda3/envs/p810/bin/flask", line 11, in <module>
    sys.exit(main())
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask/cli.py", line 966, in main
    cli.main(prog_name="python -m flask" if as_module else None)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask/cli.py", line 586, in main
    return super(FlaskGroup, self).main(*args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 717, in main
    rv = self.invoke(ctx)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 1137, in invoke
    return _process_result(sub_ctx.command.invoke(sub_ctx))
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 956, in invoke
    return ctx.invoke(self.callback, **ctx.params)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/decorators.py", line 17, in new_func
    return f(get_current_context(), *args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask/cli.py", line 426, in decorator
    return __ctx.invoke(f, *args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/click/core.py", line 555, in invoke
    return callback(*args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask_migrate/cli.py", line 90, in migrate
    rev_id, x_arg)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask_migrate/__init__.py", line 95, in wrapped
    f(*args, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/flask_migrate/__init__.py", line 215, in migrate
    version_path=version_path, rev_id=rev_id)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/alembic/command.py", line 214, in revision
    script_directory.run_env()
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/alembic/script/base.py", line 489, in run_env
    util.load_python_file(self.dir, "env.py")
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/alembic/util/pyfiles.py", line 98, in load_python_file
    module = load_module_py(module_id, path)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/alembic/util/compat.py", line 184, in load_module_py
    spec.loader.exec_module(module)
  File "<frozen importlib._bootstrap_external>", line 697, in exec_module
  File "<frozen importlib._bootstrap>", line 222, in _call_with_frames_removed
  File "migrations/env.py", line 96, in <module>
    run_migrations_online()
  File "migrations/env.py", line 81, in run_migrations_online
    with connectable.connect() as connection:
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 2206, in connect
    return self._connection_cls(self, **kwargs)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 103, in __init__
    else engine.raw_connection()
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 2306, in raw_connection
    self.pool.unique_connection, _connection
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 2279, in _wrap_pool_connect
    e, dialect, self
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 1544, in _handle_dbapi_exception_noconnection
    util.raise_from_cause(sqlalchemy_exception, exc_info)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 398, in raise_from_cause
    reraise(type(exception), exception, tb=exc_tb, cause=cause)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/util/compat.py", line 152, in reraise
    raise value.with_traceback(tb)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/base.py", line 2275, in _wrap_pool_connect
    return fn()
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 304, in unique_connection
    return _ConnectionFairy._checkout(self)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 761, in _checkout
    fairy = _ConnectionRecord.checkout(pool)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 493, in checkout
    rec = pool._do_get()
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/impl.py", line 238, in _do_get
    return self._create_connection()
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 309, in _create_connection
    return _ConnectionRecord(self)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 438, in __init__
    self.__connect(first_connect_check=True)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/pool/base.py", line 640, in __connect
    connection = pool._invoke_creator(self)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/strategies.py", line 115, in connect
    return dialect.connect(*cargs, **cparams)
  File "/anaconda3/envs/p810/lib/python3.5/site-packages/sqlalchemy/engine/default.py", line 483, in connect
    return self.dbapi.connect(*cargs, **cparams)
sqlalchemy.exc.OperationalError: (pyodbc.OperationalError) ('08001', '[08001] [Microsoft][ODBC Driver 17 for SQL Server]Neither DSN nor SERVER keyword supplied (0) (SQLDriverConnect)')`

I've tried to debug and can only find that DefaultEngineStrategy.create is being called twice.  In the first attempt, all seems well when name_or_url is <class 'sqlalchemy.engine.url.URL'>.  On the second attempt, it is a string, and make_url() returns:  

mssql+pyodbc:///?Connection Timeout=30&Database=dbname&Encrypt=yes&Pwd=mypword&Server=servername&TrustServerCertificate=no&Uid=username&odbc_connect=Driver={ODBC Driver 17 for SQL Server}

which causes `(cargs, cparams) = dialect.create_connect_args(u)` to produce:  `[['Driver={ODBC Driver 17 for SQL Server}'], {}]`

Any help is appreciated.

**Versions.**
 - OS: macOS Mojave
 - Python: 3.5
 - SQLAlchemy:  1.3.10
 - Database:  MS SQL Server

**Comment by zzzeek** [comment]
hi there -

If i use SQLAlchemy directly, the code example you gave doesn't produce that effect on this end nor can I see any reason why it should.  So this must be something on the Flask side.

your code as given creates a good URL:

```
>>> import urllib.parse
>>> params = urllib.parse.quote_plus \
... ("Driver={ODBC Driver 17 for SQL Server};Server=servername;Database=dbname;Uid=username;Pwd=mypword;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;")
>>> conn_str = 'mssql+pyodbc:///?odbc_connect={}'.format(params)
>>> from sqlalchemy.engine.url import make_url
>>> make_url(conn_str)
mssql+pyodbc:///?odbc_connect=Driver={ODBC Driver 17 for SQL Server};Server=servername;Database=dbname;Uid=username;Pwd=mypword;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;

```

create_connect_args() makes good arguments also which don't look like the ones you are seeing:

```
>>> from sqlalchemy.dialects.mssql import pyodbc
>>> dialect = pyodbc.dialect()
>>> url = make_url(conn_str)
>>> dialect.create_connect_args(url)
[['Driver={ODBC Driver 17 for SQL Server};Server=servername;Database=dbname;Uid=username;Pwd=mypword;Encrypt=yes;TrustServerCertificate=no;Connection Timeout=30;'], {}]

```

I'm not sure what flask-migrate does with SQLALCHEMY_DATABASE_URI but perhaps it is getting mis-handled within the exec / click process it seems to be using.

Place your URL logic directly in your Alembic env.py to confirm there's no issue when you bypass the Flask mechanics.

**Comment by tomashornak** [comment]
Hi.
I was able to find workaround for this issue. Check my comment here: https://github.com/miguelgrinberg/Flask-Migrate/issues/328#issuecomment-629870874

Anyway. To me this seems like Flask-Migrate (maybe Alembic?)

**Comment by tomas223** [comment]
Hi.
I was able to find workaround for this issue. Check my comment here: https://github.com/miguelgrinberg/Flask-Migrate/issues/328#issuecomment-629870874

Anyway. To me this seems like Flask-Migrate (maybe Alembic?)

**Comment by clabornd** [comment]
Thanks for the responses, creating the connection string in env.py worked.  I suppose this can be closed as it appears to not be a sqlalchemy issue, though the Flask-Migrate maintainer suggests it is.

Funnily enough, I now cannot replicate the double call to DefaultEngineStrategy.create...

**Comment by zzzeek** [comment]
@clabornd  @tomas223 im a little puzzled by the two identities here, are you working togehter on the same issue?  in any case it's suggested in https://github.com/miguelgrinberg/Flask-Migrate/issues/328#issuecomment-630257975 that this has been fixed, so perhaps if you have a generated env.py that predates this fix then amending the file is the correct solution to the problem.

**Comment by zzzeek** [comment]
the URL issue that's been identified will be fixed with #5341
